### PR TITLE
Implement CallProfiler

### PR DESCRIPTION
**Pull request by tigercosmos**
This work is based on the previous #245

This work introduces a new profiler, called `CallProfiler`, which can print the profiling result of a hierarchical caller stack.

The example output looks like this (please refer to `test_nopython_callprofiler.cpp`):

```c++
constexpr int uniqueTime1 = 19;
constexpr int uniqueTime2 = 35;
constexpr int uniqueTime3 = 7;

void foo3()
{
    modmesh::USE_CALLPROFILER_PROFILE_THIS_FUNCTION();
    std::this_thread::sleep_for(std::chrono::milliseconds(uniqueTime1));
}

void foo2()
{
    modmesh::USE_CALLPROFILER_PROFILE_THIS_FUNCTION();
    std::this_thread::sleep_for(std::chrono::milliseconds(uniqueTime2));
    foo3();
}

void foo1()
{
    modmesh::USE_CALLPROFILER_PROFILE_THIS_FUNCTION();
    foo2();
    std::this_thread::sleep_for(std::chrono::milliseconds(uniqueTime3));
}


int main()
{
    auto & profiler = modmesh::Profiler::instance();

    foo1();
    foo2();
    foo3();
    foo3();
  
    profiler.print_profiling_result();
}
```

```log
Profiling Result
  foo1 - Total Time: 61 ms, Call Count: 1
    foo2 - Total Time: 54 ms, Call Count: 1
      foo3 - Total Time: 19 ms, Call Count: 1
  foo2 - Total Time: 54 ms, Call Count: 1
    foo3 - Total Time: 19 ms, Call Count: 1
  foo3 - Total Time: 38 ms, Call Count: 2
```

cc @q40603

### CODE REVIEW

File: gtests/test_nopython_callprofiler.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    mm::CallProfiler::instance();
+}
+
+constexpr int uniqueTime1 = 19;
+constexpr int uniqueTime2 = 35;
+constexpr int uniqueTime3 = 7;
+
+void foo3()
+{
+    modmesh::USE_CALLPROFILER_PROFILE_THIS_FUNCTION();
```
Comment (tigercosmos) [inline gtests/test_nopython_callprofiler.cpp:21]: we use the macro here to insert the profiler.

**Comment by tigercosmos** [comment]
The implementation is almost there. I will fix the CI once the proposal is accepted.

File: gtests/test_nopython_callprofiler.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    std::stringstream ss;
+    profiler.print_profiling_result(ss);
+
+    const char * answer = R"(Profiling Result
+  foo1 - Total Time: 61 ms, Call Count: 1
+    foo2 - Total Time: 54 ms, Call Count: 1
+      foo3 - Total Time: 19 ms, Call Count: 1
+)";
+
+    EXPECT_EQ(ss.str(), answer);
```
Comment (tigercosmos) [inline gtests/test_nopython_callprofiler.cpp:57]: just for demonstration here, I think it's better to compare each node one by one.

**Comment by yungyuc** [comment]
> The implementation is almost there. I will fix the CI once the proposal is accepted.

Let me add comments after the feedback from @q40603 .

But I would like the code review to happen _after_ CI is clean.  Fixing CI needs to change code, which requires additional review.  It does not make sense to ask for review before making CI clean.

modmesh CI includes linter.  It wastes precious reviewing resources when the linter automatically tells you there are things needing to be fixed.

**Comment by tigercosmos** [comment]
I would like to have more discussion on this PR, let me close this one can create a new draft PR.

**Pull request closed without merging**